DashboardTest.testComponent_linkageCheckResult fails in Java 11

DashboardTest.testComponent_linkageCheckResult fails in Java 11 as per https://github.com/GoogleCloudPlatform/cloud-opensource-java/runs/1481215441.

```
2020-12-01T16:48:46.5244605Z [ERROR] Tests run: 22, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 59.252 s <<< FAILURE! - in com.google.cloud.tools.opensource.dashboard.DashboardTest
2020-12-01T16:48:46.5256922Z [ERROR] testComponent_linkageCheckResult(com.google.cloud.tools.opensource.dashboard.DashboardTest)  Time elapsed: 0.039 s  <<< FAILURE!
2020-12-01T16:48:46.5263327Z com.google.common.truth.ComparisonFailureWithFacts: 
2020-12-01T16:48:46.5267950Z value of: trimAndCollapseWhiteSpace(...)
2020-12-01T16:48:46.5275870Z expected: 100 target classes causing linkage errors referenced from 540 source classes.
2020-12-01T16:48:46.5277003Z but was : 105 target classes causing linkage errors referenced from 562 source classes.
2020-12-01T16:48:46.5279621Z 	at com.google.cloud.tools.opensource.dashboard.DashboardTest.testComponent_linkageCheckResult(DashboardTest.java:303)
```



# Previous Failure (resolved)

cloud-opensource-java/dashboard does not compile in Java 11 in my laptop. Why? https://gist.github.com/suztomo/a2bc02befa7195b0e9e129fb7029e3ab

```
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ dashboard ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 6 source files to /Users/suztomo/cloud-opensource-java/dashboard/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /Users/suztomo/cloud-opensource-java/dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java:[72,29] error: package org.apache.commons.cli does not exist
[ERROR] /Users/suztomo/cloud-opensource-java/dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java:[73,25] error: package org.eclipse.aether does not exist
[ERROR] /Users/suztomo/cloud-opensource-java/dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java:[74,25] error: package org.eclipse.aether does not exist
[ERROR] /Users/suztomo/cloud-opensource-java/dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java:[75,34] error: package org.eclipse.aether.artifact does not exist
[ERROR] /Users/suztomo/cloud-opensource-java/dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java:[76,34] error: package org.eclipse.aether.artifact does not exist
[ERROR] /Users/suztomo/cloud-opensource-java/dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java:[77,31] error: package org.eclipse.aether.graph does not exist
[ERROR] /Users/suztomo/cloud-opensource-java/dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java:[101,45] error: cannot find symbol
  symbol:   class RepositoryException
  location: class DashboardMain
```

> [WARNING] The POM for com.google.cloud.tools:dependencies:jar:1.5.5-SNAPSHOT is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details

Maybe the parent pon.xml was not installed beforehand.


# Identifying the new linkage error that causes the difference

Step 1: Run LinkageCheckerMain with `-b com.google.cloud:libraries-bom:16.1.0 -o /tmp/baseline.txt` option on Java 8:

<img width="1065" alt="Screen Shot 2020-12-02 at 10 53 53" src="https://user-images.githubusercontent.com/28604/100896604-ca5b9f00-348c-11eb-8a34-d40ced5a87cb.png">

Step 2: Run LinkageCheckerMain with -b com.google.cloud:libraries-bom:16.1.0 -e /tmp/baseline.txt` option on Java 11.

<img width="1053" alt="Screen Shot 2020-12-02 at 10 55 27" src="https://user-images.githubusercontent.com/28604/100896785-00008800-348d-11eb-9451-2e46bb161fae.png">

The difference were `javax.activation.DataSource`, `javax.activation.FileDataSource`,  `javax.activation.DataHandler`, `javax.activation.ActivationDataFlavor`, `javax.activation.DataContentHandler`

```
Class javax.activation.DataSource is not found;
  referenced by 9 class files
    com.google.appengine.api.utils.HttpRequestParser (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    com.google.apphosting.utils.servlet.MultipartMimeUtils (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    javax.mail.MultipartDataSource (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    javax.mail.internet.MimeMultipart (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    javax.mail.internet.MimePartDataSource (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    javax.mail.internet.MimeUtility (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    javax.mail.util.ByteArrayDataSource (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    org.apache.geronimo.mail.handlers.MessageHandler (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    org.apache.geronimo.mail.handlers.TextHandler (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
  Cause:
    Unknown
Class javax.activation.FileDataSource is not found;
  referenced by 1 class file
    javax.mail.internet.MimeBodyPart (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
  Cause:
    Unknown
Class javax.activation.DataHandler is not found;
  referenced by 4 class files
    javax.mail.internet.MimeBodyPart (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    javax.mail.internet.MimeMessage (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    javax.mail.internet.MimeUtility (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    javax.mail.internet.PreencodedMimeBodyPart (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
  Cause:
    Unknown
Class javax.activation.ActivationDataFlavor is not found;
  referenced by 5 class files
    org.apache.geronimo.mail.handlers.HtmlHandler (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    org.apache.geronimo.mail.handlers.MessageHandler (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    org.apache.geronimo.mail.handlers.MultipartHandler (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    org.apache.geronimo.mail.handlers.TextHandler (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    org.apache.geronimo.mail.handlers.XMLHandler (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
  Cause:
    Unknown
Class javax.activation.DataContentHandler is not found;
  referenced by 3 class files
    org.apache.geronimo.mail.handlers.MessageHandler (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    org.apache.geronimo.mail.handlers.MultipartHandler (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
    org.apache.geronimo.mail.handlers.TextHandler (com.google.appengine:appengine-api-1.0-sdk:1.9.71)
  Cause:
    Unknown
Problematic artifacts in the dependency tree:
com.google.appengine:appengine-api-1.0-sdk:1.9.71 is at:
  com.google.http-client:google-http-client-appengine:1.38.0 (compile) / com.google.appengine:appengine-api-1.0-sdk:1.9.71 (provided)
  and 2 other dependency paths.
```

`javax.activation` package has been removed in Java 11 ([JEP 320: Remove the Java EE and CORBA Modules](https://openjdk.java.net/jeps/320))

--- a/dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java
+++ b/dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java
@@ -48,8 +48,10 @@ import nu.xom.XPathContext;
 import org.eclipse.aether.artifact.Artifact;
 import org.eclipse.aether.artifact.DefaultArtifact;
 import org.eclipse.aether.resolution.ArtifactDescriptorException;
+import org.hamcrest.core.StringStartsWith;
 import org.junit.AfterClass;
 import org.junit.Assert;
+import org.junit.Assume;
 import org.junit.BeforeClass;
 import org.junit.Test;
 
@@ -293,7 +295,8 @@ public class DashboardTest {
   }
 
   @Test
-  public void testComponent_linkageCheckResult() throws IOException, ParsingException {
+  public void testComponent_linkageCheckResult_java8() throws IOException, ParsingException {
+    Assume.assumeThat(System.getProperty("java.version"), StringStartsWith.startsWith("1.8."));
     // version used in libraries-bom 1.0.0
     Document document = parseOutputFile(
         "com.google.http-client_google-http-client-appengine_1.29.1.html");
@@ -307,6 +310,34 @@ public class DashboardTest {
     Nodes causes = document.query("//p[@class='jar-linkage-report-cause']");
     Truth.assertWithMessage(
             "google-http-client-appengine should show linkage errors for RpcStubDescriptor")
+        .that(causes)
+        .comparingElementsUsing(NODE_VALUES)
+        .contains(
+            "Class com.google.net.rpc3.client.RpcStubDescriptor is not found,"
+                + " referenced from 21 classes ▶"); // '▶' is the toggle button
+  }
+
+  @Test
+  public void testComponent_linkageCheckResult_java11() throws IOException, ParsingException {
+    String javaVersion = System.getProperty("java.version");
+    // javaMajorVersion is 1 when we use Java 8. Still good indicator to ensure Java 11 or higher.
+    int javaMajorVersion = Integer.parseInt(javaVersion.split("\\.")[0]);
+    Assume.assumeTrue(javaMajorVersion >= 11);
+
+    // version used in libraries-bom 1.0.0
+    Document document = parseOutputFile(
+        "com.google.http-client_google-http-client-appengine_1.29.1.html");
+    Nodes reports = document.query("//p[@class='jar-linkage-report']");
+    Assert.assertEquals(2, reports.size());
+    // This number differs between Java 8 and Java 11
+    Truth.assertThat(trimAndCollapseWhiteSpace(reports.get(0).getValue()))
+        .isEqualTo("105 target classes causing linkage errors referenced from 562 source classes.");
+    Truth.assertThat(trimAndCollapseWhiteSpace(reports.get(1).getValue()))
+        .isEqualTo("3 target classes causing linkage errors referenced from 3 source classes.");
+
+    Nodes causes = document.query("//p[@class='jar-linkage-report-cause']");
+    Truth.assertWithMessage(
+        "google-http-client-appengine should show linkage errors for RpcStubDescriptor")
         .that(causes)
         .comparingElementsUsing(NODE_VALUES)
         .contains(

--- a/dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java
+++ b/dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java
@@ -297,7 +297,7 @@ public class DashboardTest {
   @Test
   public void testComponent_linkageCheckResult_java8() throws IOException, ParsingException {
     Assume.assumeThat(System.getProperty("java.version"), StringStartsWith.startsWith("1.8."));
-    // version used in libraries-bom 1.0.0
+    // The version used in libraries-bom 1.0.0
     Document document = parseOutputFile(
         "com.google.http-client_google-http-client-appengine_1.29.1.html");
     Nodes reports = document.query("//p[@class='jar-linkage-report']");
@@ -324,12 +324,17 @@ public class DashboardTest {
     int javaMajorVersion = Integer.parseInt(javaVersion.split("\\.")[0]);
     Assume.assumeTrue(javaMajorVersion >= 11);
 
-    // version used in libraries-bom 1.0.0
+    // The version used in libraries-bom 1.0.0. The google-http-client-appengine has been known to
+    // have linkage errors in its dependencies ("commons-logging:commons-logging:1.2" and
+    // "com.google.appengine:appengine-api-1.0-sdk:1.9.71").
     Document document = parseOutputFile(
         "com.google.http-client_google-http-client-appengine_1.29.1.html");
     Nodes reports = document.query("//p[@class='jar-linkage-report']");
     Assert.assertEquals(2, reports.size());
-    // This number differs between Java 8 and Java 11
+
+    // This number of linkage errors differs between Java 8 and Java 11 for the javax.activation
+    // package removal (JEP 320: Remove the Java EE and CORBA Modules). For the detail, see
+    // https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1849.
     Truth.assertThat(trimAndCollapseWhiteSpace(reports.get(0).getValue()))
         .isEqualTo("105 target classes causing linkage errors referenced from 562 source classes.");
     Truth.assertThat(trimAndCollapseWhiteSpace(reports.get(1).getValue()))

